Grammar/python.gram

```diff
@@ -394,14 +394,14 @@ for_stmt[stmt_ty]:
 
 with_stmt[stmt_ty]:
     | invalid_with_stmt_indent
-    | 'with' '(' a[asdl_withitem_seq*]=','.with_item+ ','? ')' ':' tc=[TYPE_COMMENT] b=block {
-       _PyAST_With(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA) }
-    | 'with' a[asdl_withitem_seq*]=','.with_item+ ':' tc=[TYPE_COMMENT] b=block {
-        _PyAST_With(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA) }
-    | 'async' 'with' '(' a[asdl_withitem_seq*]=','.with_item+ ','? ')' ':' b=block {
-       CHECK_VERSION(stmt_ty, 5, "Async with statements are", _PyAST_AsyncWith(a, b, NULL, EXTRA)) }
-    | 'async' 'with' a[asdl_withitem_seq*]=','.with_item+ ':' tc=[TYPE_COMMENT] b=block {
-       CHECK_VERSION(stmt_ty, 5, "Async with statements are", _PyAST_AsyncWith(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA)) }
+    | start='with' '(' a[asdl_withitem_seq*]=','.with_item+ ','? ')' end=':' tc=[TYPE_COMMENT] b=block {
+       _PyAST_With(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA_EXPR(start, end)) }
+    | start='with' a[asdl_withitem_seq*]=','.with_item+ end=':' tc=[TYPE_COMMENT] b=block {
+        _PyAST_With(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA_EXPR(start, end)) }
+    | start='async' 'with' '(' a[asdl_withitem_seq*]=','.with_item+ ','? ')' end=':' b=block {
+       CHECK_VERSION(stmt_ty, 5, "Async with statements are", _PyAST_AsyncWith(a, b, NULL, EXTRA_EXPR(start, end))) }
+    | start='async' 'with' a[asdl_withitem_seq*]=','.with_item+ end=':' tc=[TYPE_COMMENT] b=block {
+       CHECK_VERSION(stmt_ty, 5, "Async with statements are", _PyAST_AsyncWith(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA_EXPR(start, end))) }
```

Comment:
If I understand correctly, this will create a situation where the source location of a `with` node does not cover all the source code of the `body` sub-tree hanging under it. I think in the past we considered this undesirable.

If you think this is ok we can go with it, but I assumed it was not, and my intention was that we add a `header` child to a `with` node, so the tree has  a `with` node spanning header + body, with two sub-trees, one for header and one for body.  The other compound statements can also have a header + body split.

Reply:
> If I understand correctly, this will create a situation where the source location of a `with` node does not cover all the source code of the `body` sub-tree hanging under it. I think in the past we considered this undesirable.

Hummmm, I am confused. You mentioned in the issue that:

> If __enter__ or __exit__ of the context manager raises an exception, we don't want the location of the exception to span the entire body

With that I though the point was not to show the body no?

Reply:
> If I understand correctly, this will create a situation where the source location of a `with` node does not cover all the source code of the `body` sub-tree hanging under it. I think in the past we considered this undesirable.
> 
> If you think this is ok we can go with it, but I assumed it was not, and my intention was that we add a `header` child to a `with` node, so the tree has a `with` node spanning header + body, with two sub-trees, one for header and one for body. The other compound statements can also have a header + body split.

Ah, that is actually a breaking change for tools :(

Reply:
> If you think this is ok we can go with it, but I assumed it was not, and my intention was that we add a `header` child to a `with` node, so the tree has a `with` node spanning header + body, with two sub-trees, one for header and one for body. The other compound statements can also have a header + body split.

Yeah, I can go with this instead :+1:

Reply:
Just to be clear, how different are you picturing this to be from the current version. Currently for

```
with A, B, C:
    ...
```

we have:

```
Module(
   body=[
      With(
         items=[
            withitem(
               context_expr=Name(id='A', ctx=Load())),
            withitem(
               context_expr=Name(id='B', ctx=Load())),
            withitem(
               context_expr=Name(id='C', ctx=Load()))],
         body=[
            Expr(
               value=Constant(value=Ellipsis))])])
```

are you picturing some new node wrapping the items?

Reply:
I have made a version of this just adding a "header" wrapper

Reply:
I think we need the location of the whole with statement, up to the `:`. So it would need a new node wrapping all the items. 

But let's make sure we have a plan. How do we want to render a traceback for an exception coming out of the `__enter__` or `__exit__` of one of the context managers? Should we hilight the whole statement and put carets under the expression that creates the specific context manager that caused the error?

CC @ericsnowcurrently